### PR TITLE
It's not a mixer (2/3)

### PR DESCRIPTION
**Pull request by pmikolajczyk41**
We add `withdraw` method.

### CODE REVIEW

File: contracts/blender/src/contract.rs [hunk 1]
```diff
@@ -33,7 +34,15 @@ mod blender {
...
-        note: Note,
+        leaf_idx: u32,
+    }
+
+    #[ink(event)]
+    pub struct Withdrawn {
+        #[ink(topic)]
+        token_id: TokenId,
+        value: TokenAmount,
+        recipient: AccountId,
```
Comment (deuszx) [inline contracts/blender/src/contract.rs:45]: we could index recipient as well.

Comment (pmikolajczyk41) [reply contracts/blender/src/contract.rs:46]: done

File: contracts/blender/src/contract.rs [hunk 2]
```diff
@@ -82,22 +94,114 @@ mod blender {
...
+            self.verify_merkle_root(merkle_root)?;
+            self.verify_nullifier(nullifier)?;
+            self.verify_withdrawal(token_id, value, merkle_root, nullifier, new_note, proof)?;
+
+            let leaf_idx = self.append_note(new_note)?;
+            self.nullifiers.insert(nullifier, &());
+
+            self.withdraw_funds(token_id, value, fee_for_caller, recipient)?;
+
+            self.env().emit_event(Withdrawn {
```
Comment (deuszx) [inline contracts/blender/src/contract.rs:133]: I suggest using the approach described [here](https://github.com/deuszx/bulletin-board-example/blob/d6f00049f6c07a113d0e46b6c66825f3f382f236/contracts/README.md#creating-and-emitting-events).

Comment (pmikolajczyk41) [reply contracts/blender/src/contract.rs]: done

File: contracts/blender/src/contract.rs [hunk 3]
```diff
@@ -185,41 +287,98 @@ mod blender {
...
+            }
+        }
+
+        fn transfer(
+            &self,
+            token_contract: AccountId,
+            value: TokenAmount,
+            recipient: AccountId,
+        ) -> Result<()> {
+            build_call::<super::blender::Environment>()
```
Comment (deuszx) [inline contracts/blender/src/contract.rs:371]: You could use the *Ref pattern here instead. Calls this simple are supported.

**Pull request merged**